pkg/webhook/resources/user/user.go

```diff
@@ -0,0 +1,109 @@
+package user
+
+import (
+	"crypto/sha256"
+	"encoding/base32"
+	"fmt"
+	"strings"
+
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/harvester/harvester/pkg/apis/harvesterhci.io"
+	"github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	ctlharvesterv1 "github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1"
+	"github.com/harvester/harvester/pkg/indexeres"
+	pkguser "github.com/harvester/harvester/pkg/user"
+	werror "github.com/harvester/harvester/pkg/webhook/error"
+	"github.com/harvester/harvester/pkg/webhook/types"
+)
+
+func NewMutator(users ctlharvesterv1.UserCache) types.Mutator {
+	users.AddIndexer(indexeres.UserNameIndex, indexeres.IndexUserByUsername)
+	return &userMutator{users: users}
+}
+
+type userMutator struct {
+	types.DefaultMutator
+
+	users ctlharvesterv1.UserCache
+}
+
+func (m *userMutator) Resource() types.Resource {
+	return types.Resource{
+		Name:       v1beta1.UserResourceName,
+		Scope:      admissionregv1.ClusterScope,
+		APIGroup:   harvesterhci.GroupName,
+		APIVersion: v1beta1.SchemeGroupVersion.Version,
+		ObjectType: &v1beta1.User{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+			admissionregv1.Delete,
+		},
+	}
+}
+
+func (m *userMutator) Create(request *types.Request, newObj runtime.Object) (types.PatchOps, error) {
+	user := newObj.(*v1beta1.User)
+	return m.createOrUpdateUser(user, true)
+}
+
+func (m *userMutator) Update(request *types.Request, oldObj runtime.Object, newObj runtime.Object) (types.PatchOps, error) {
+	user := newObj.(*v1beta1.User)
+	return m.createOrUpdateUser(user, false)
+}
+
+func (m *userMutator) Delete(request *types.Request, oldObj runtime.Object) (types.PatchOps, error) {
+	user := oldObj.(*v1beta1.User)
+	if user.Name == request.AdmissionRequest.UserInfo.Username {
+		return nil, werror.NewInvalidError("can't delete self", "metadata.name")
+	}
```

Comment:
Should we separate the validation stuff to validating webhook?

Reply:
Agree, [separate to validation](https://github.com/harvester/harvester/pull/919/commits/42f8234f9f3ad78a6d070b89848a926d89a84dce#diff-f861497ca56a5f68d92c9f4bd04f1d929d6dec18950628d560bb78cb5711e6d1).

Reply:
The validation for two mutators are made individually.
- 1d45ec2
- b757824